### ISSUE TITLE
Policy controller doesn't support mixed IP SANs/DNS SANs

### ISSUE THREAD

**Issue by fasaxc** [projectcalico/k8s-policy#69]
This often trips people up, especially if they're following a perfectly reasonable guide to get k8s certs set up then they try to add Calico.

I believe IP SAN support has now landed in urllib3 so revving that library might get us that support "for free".

**Comment by fasaxc** [comment]
Assigning to Casey for triage...

**Comment by caseydavenport** [comment]
This was fixed in PR https://github.com/projectcalico/k8s-policy/pull/48.

@fasaxc what version are you seeing this in?

**Comment by fasaxc** [comment]
It was reported by **kitch** on slack, not sure of the version.

**Comment by fasaxc** [comment]
They're using 0.5.2

**Comment by kitch** [comment]
Right, so the trick is this. We are running our workers behind a proxy. So master is public, and worker is walled off. In order to allow users inside the wall and outside the wall to use this beast, we use a domain name for the kube-apiserver to contact it. So thus the SAN containers IP and DNS..

```
X509v3 Subject Alternative Name:
IP Address:10.143.139.206, IP Address:10.143.139.239, IP Address:10.143.139.211, IP Address:10.143.139.212, IP Address:10.143.139.234, IP Address:10.143.139.215, IP Address:10.10.10.1, DNS:api-stage.cont.XXXXXX.net
```

10.10.10.1 is needed because all pods use this address to talk to kube api-server, outside of kube we use the dns name to talk to kube.

**Comment by treytabner** [comment]
To clarify the problem here, when we have both IP addresses and one or more DNS names listed in the certificate, the IP addresses are ignored and only the DNS name is checked:

```
2017-03-31 16:16:20,100 5 INFO Syncing 'Namespace' objects
2017-03-31 16:16:20,100 5 DEBUG Getting API resources 'https://172.20.0.1:443/api/v1/namespaces' at version 'None'. stream=False
2017-03-31 16:16:20,100 5 INFO Syncing 'Pod' objects
2017-03-31 16:16:20,102 5 DEBUG Getting API resources 'https://172.20.0.1:443/api/v1/pods' at version 'None'. stream=False
2017-03-31 16:16:20,104 5 INFO Syncing 'NetworkPolicy' objects
2017-03-31 16:16:20,104 5 DEBUG Getting API resources 'https://172.20.0.1:443/apis/extensions/v1beta1/networkpolicies' at version 'None'. stream=False
2017-03-31 16:16:20,193 5 WARNING Connection error: Pod ...hostname '172.20.0.1' doesn't match 'dev.cont.example.net'
2017-03-31 16:16:20,194 5 WARNING Connection error: Namespace ...hostname '172.20.0.1' doesn't match 'dev.cont.example.net'
2017-03-31 16:16:20,195 5 INFO Restarting watch on resource: Namespace
2017-03-31 16:16:20,195 5 WARNING Connection error: NetworkPolicy ...hostname '172.20.0.1' doesn't match 'dev.cont.example.net'
2017-03-31 16:16:20,195 5 INFO Restarting watch on resource: NetworkPolicy
2017-03-31 16:16:20,195 5 INFO Restarting watch on resource: Pod
```

The certificate in question contains:

```
            X509v3 Subject Alternative Name: 
                IP Address:192.168.10.2, IP Address:192.168.10.3, IP Address:192.168.10.4, IP Address:172.20.0.1, DNS:dev.cont.example.net
```

**Comment by kitch** [comment]
From ssl.py

```
    dnsnames = []
    san = cert.get('subjectAltName', ())
    for key, value in san:
        if key == 'DNS':
            if _dnsname_match(value, hostname):
                return
            dnsnames.append(value)
    if not dnsnames:
        # The subject is only checked when there is no dNSName entry
        # in subjectAltName
        for sub in cert.get('subject', ()):
            for key, value in sub:
                # XXX according to RFC 2818, the most specific Common Name
                # must be used.
                if key == 'commonName':
                    if _dnsname_match(value, hostname):
                        return
                    dnsnames.append(value)
```

As you can see, python ignores the IP SANs if there are ANY DNS SANs in the cert.

**Comment by djosborne** [comment]
That's the correct behavior of an SSL client, [as per digicert](https://www.digicert.com/subject-alternative-name-compatibility.htm):

>If a SSL Certificate has a Subject Alternative Name (SAN) field, then SSL clients are supposed to ignore the Common Name value and seek a match in the SAN list. This is why DigiCert always repeats the common name as the first SAN in our certificates.

**Comment by fasaxc** [comment]
The libcalico change to rev requests is now merged.  Just need to pull that into a policy controller release.

**Comment by tomdee** [comment]
Fixed in 0.6.0 https://github.com/projectcalico/k8s-policy/releases/tag/v0.6.0
